### PR TITLE
[6.0] New collection methods

### LINKED ISSUE #1004: New collection methods for 6.0.0
Methods which were proposed but will not be implemented: 
* `zipWith` (https://github.com/ScorexFoundation/sigmastate-interpreter/issues/419) as it can be replaced with `zip` and `map` , for example, instead of `def add(c1: Coll[Byte], c2: Coll[Byte]): Coll[Byte] = c1.zipWith(c2, { (x, y) => x + y })` we can do smth like `
def add(c1: Coll[Byte], c2: Coll[Byte]): Coll[Byte] = c1.zip(c2).map({ (c: (Byte, Byte)) => (c._1 + c._2).toByte })`, and we can do example from #419 once bitwise `^` is supported

### PR DESCRIPTION
**Pull request by kushti**
In this PR, new collection methods got introduced:

* reverse
* distinct
* startsWith
* endsWith
* get

close #1004

### CODE REVIEW

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala [hunk 1]
```diff
@@ -1611,4 +1611,94 @@ class LanguageSpecificationV6 extends LanguageSpecificationBase { suite =>
...
+      { (xs: Coll[Int]) => xs.distinct },
+      """{(xs: Coll[Int]) => xs.distinct }""".stripMargin,
+      sinceVersion = VersionContext.V6SoftForkVersion
+    )
+
+    verifyCases(
+      Seq(
+        Coll(1, 2) -> Expected(ExpectedResult(Success(Coll(1, 2)), None)),
+        Coll(1, 1, 2) -> Expected(ExpectedResult(Success(Coll(1, 2)), None)),
+        Coll(1, 2, 2) -> Expected(ExpectedResult(Success(Coll(1, 2)), None)),
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:1641]: Suggest to add a case with unordered collections and many different duplicates.

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala [hunk 2]
```diff
@@ -1611,4 +1611,94 @@ class LanguageSpecificationV6 extends LanguageSpecificationBase { suite =>
...
+      """{(xs: (Coll[Int], Coll[Int])) => xs._1.startsWith(xs._2) }""".stripMargin,
+      sinceVersion = VersionContext.V6SoftForkVersion
+    )
+
+    verifyCases(
+      Seq(
+        (Coll(1, 2, 3), Coll(1, 2)) -> Expected(ExpectedResult(Success(true), None)),
+        (Coll(1, 2, 3), Coll(1, 2, 3)) -> Expected(ExpectedResult(Success(true), None)),
+        (Coll(1, 2, 3), Coll(1, 2, 3, 4)) -> Expected(ExpectedResult(Success(false), None)),
+        (Coll[Int](), Coll[Int]()) -> Expected(ExpectedResult(Success(true), None))
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:1660]: Not covered cases:
- Coll(1, 2, 3), Coll(1, 2, 4)
- Coll[Int](1, 2), Coll[Int]()

**Review by aslesarenko — APPROVED** [review]
Approved, but please address the comments.

File: core/shared/src/main/scala/sigma/Colls.scala [hunk 3]
```diff
@@ -45,6 +45,22 @@ trait Coll[@specialized A] {
...
 
+  /** The element at given index.
+    *  Indices start at `0`; `xs.apply(0)` is the first element of collection `xs`.
+    *  Note the indexing syntax `xs(i)` is a shorthand for `xs.apply(i)`.
+    *
+    *  @param  i       the index
+    *  @return         the element at the given index
+    *  @throws         ArrayIndexOutOfBoundsException if `i < 0` or `length <= i`
+    */
+  def get(i: Int): Option[A] = {
```
[the remark this answers is not in the document]
Comment (kushti) [reply core/shared/src/main/scala/sigma/Colls.scala:56]: fixed

File: core/shared/src/main/scala/sigma/Colls.scala [hunk 4]
```diff
@@ -76,6 +92,14 @@ trait Coll[@specialized A] {
     * produces a collection ((x0, y0), ..., (xK, yK)) where K = min(N, M) */
   def zip[@specialized B](ys: Coll[B]): Coll[(A, B)]
 
+  /** For this collection (x0, ..., xN) and other collection (y0, ..., yM)
+    * produces a collection ((x0, y0), ..., (xK, yK)) where K = min(N, M) */
+  def startsWith(ys: Coll[A]): Boolean
```
[the remark this answers is not in the document]
Comment (kushti) [reply core/shared/src/main/scala/sigma/Colls.scala:97]: fixed

File: core/shared/src/main/scala/sigma/Colls.scala [hunk 5]
```diff
@@ -76,6 +92,14 @@ trait Coll[@specialized A] {
...
+
+  /** For this collection (x0, ..., xN) and other collection (y0, ..., yM)
+    * produces a collection ((x0, y0), ..., (xK, yK)) where K = min(N, M) */
+  def endsWith(ys: Coll[A]): Boolean
```
[the remark this answers is not in the document]
Comment (kushti) [reply core/shared/src/main/scala/sigma/Colls.scala:101]: fixed

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala (hunk 2 above)
Comment (kushti) [reply sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:1660]: added

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala (hunk 1 above)
Comment (kushti) [reply sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:1641]: added

File: core/shared/src/main/scala/sigma/data/CollsOverArrays.scala [hunk 6]
```diff
@@ -350,6 +354,10 @@ class PairOfCols[@specialized L, @specialized R](val ls: Coll[L], val rs: Coll[R
 
   def zip[@specialized B](ys: Coll[B]): PairColl[(L,R), B] = builder.pairColl(this, ys)
 
+  def startsWith(ys: Coll[(L, R)]): Boolean = toArray.startsWith(ys.toArray)
+
+  def endsWith(ys: Coll[(L, R)]): Boolean = toArray.endsWith(ys.toArray)
```
[the remark this answers is not in the document]
Comment (kushti) [reply core/shared/src/main/scala/sigma/data/CollsOverArrays.scala:359]: done, added tests also

File: data/shared/src/main/scala/sigma/ast/methods.scala [hunk 7]
```diff
@@ -1069,29 +1068,120 @@ object SCollectionMethods extends MethodsContainer with MethodByNameUnapply {
     }
   }
 
+  // ======== 6.0 methods below ===========
+
+  val ReverseMethod = SMethod(this, "reverse",
+    SFunc(Array(ThisType), ThisType, paramIVSeq), 30, Zip_CostKind) // todo: costing
```
[the remark this answers is not in the document]
Comment (kushti) [reply data/shared/src/main/scala/sigma/ast/methods.scala:1074]: fixed

File: data/shared/src/main/scala/sigma/ast/methods.scala [hunk 8]
```diff
@@ -1069,29 +1068,120 @@ object SCollectionMethods extends MethodsContainer with MethodByNameUnapply {
...
+  def reverse_eval[A](mc: MethodCall, xs: Coll[A])
+                    (implicit E: ErgoTreeEvaluator): Coll[A] = {
+    val m = mc.method
+    E.addSeqCost(m.costKind.asInstanceOf[PerItemCost], xs.length, m.opDesc) { () =>
+      xs.reverse
+    }
+  }
+
+  val DistinctMethod = SMethod(this, "distinct",
+    SFunc(Array(ThisType), ThisType, paramIVSeq), 31, Zip_CostKind) // todo: costing
```
[the remark this answers is not in the document]
Comment (kushti) [reply data/shared/src/main/scala/sigma/ast/methods.scala:1091]: fixed

**Pull request merged**